### PR TITLE
Corrections for new nogo analyzers

### PR DESCRIPTION
**Pull request by tempoz**
- Enable more Analyzers for the nogo target
- Fix broken struct tags
- Fix shallow copy of proto
- Removed unreachable code (which only existed previously because it was unreachable)
- Move defer close to after error check
- Some proto package declarations were incorrect
- Can't consume a proto channel with `range`
- fix closure capture of loop var
- Can't pass proto by value because we can't copy protos
- Error code was not being captured and thus not being checked
- Cleaned up composite literals for nogo analyzers
- Don't copy proto

---

**Version bump**: Patch

### CODE REVIEW

File: nogo_config.json [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    }
+  },
+  "unreachable": {
+    "exclude_files": {
+      "external/*": "Unreachable third party code"
+    }
+  },
+  "unsafeptr": {
+    "exclude_files": {
+      "external/*": "Unsafe third party code"
```
Comment (tylerwilliams) [inline nogo_config.json:55]: ubernit: should these just say Third party code? or Why are these analyzers different?

File: server/build_event_protocol/build_event_proxy/build_event_proxy.go [hunk 2]
```diff
@@ -80,7 +80,11 @@ func (c *BuildEventProxyClient) newAsyncStreamProxy(ctx context.Context, opts ..
 			return
 		}
 		asp.PublishBuildEvent_PublishBuildToolEventStreamClient = stream
-		for req := range asp.events {
```
Comment (tylerwilliams) [inline server/build_event_protocol/build_event_proxy/build_event_proxy.go:83]: nit: could you add a comment about this one? It's kind of interesting :) 

// Iterating over a channel copies the values in that channel, and
// copies of protos are not permitted, so rather than range over the
// channel we read from the channel inside of an outer loop.


or something like that?

File: server/build_event_protocol/event_parser/BUILD [hunk 3]
```diff
@@ -11,6 +11,7 @@ go_library(
         "//proto:invocation_go_proto",
         "//server/terminal",
         "//server/util/git",
+				"@org_golang_google_protobuf//proto:go_default_library",
```
Comment (tylerwilliams) [inline server/build_event_protocol/event_parser/BUILD:14]: might need to run buildfix on this

File: server/build_event_protocol/event_parser/event_parser_test.go [hunk 4]
```diff
@@ -204,40 +204,40 @@ func TestFillInvocation(t *testing.T) {
...
+	assert.Equal(t, 1, len(invocation.Event[10].BuildEvent.GetTestSummary().Failed))
+	assert.Equal(t, "uri", invocation.Event[10].BuildEvent.GetTestSummary().Passed[0].GetUri())
+	assert.Equal(t, "uri", invocation.Event[10].BuildEvent.GetTestSummary().Failed[0].GetUri())
 
 	assert.Equal(t, int64(1000), invocation.DurationUsec)
 
-	assert.Equal(t, "SHELL=/bin/bash", shellOption.OptionValue)
-	assert.Equal(t, "SECRET=<REDACTED>", secretOption.OptionValue)
+	assert.Equal(t, "SHELL=/bin/bash", invocation.StructuredCommandLine[0].Sections[0].GetOptionList().Option[0].OptionValue)
+	assert.Equal(t, "SECRET=<REDACTED>", invocation.StructuredCommandLine[0].Sections[0].GetOptionList().Option[1].OptionValue)
```
Comment (tylerwilliams) [inline server/build_event_protocol/event_parser/event_parser_test.go:240]: I don't fully understand why these changes are necessary. I'm not totally opposed, but they definitely make the tests harder to read / modify since you now need to think about the event index.

Did these all fail or just the commandline one due to the change above?

File: server/build_event_protocol/event_parser/event_parser.go [hunk 5]
```diff
@@ -56,8 +57,8 @@ func parseAndFilterCommandLine(in *command_line.CommandLine, allowedEnvVars []st
 	if in == nil {
 		return nil, envVarMap
 	}
-	var out command_line.CommandLine
-	out = *in
+	out := &command_line.CommandLine{}
+	proto.Merge(out, in)
```
Comment (tylerwilliams) [inline server/build_event_protocol/event_parser/event_parser.go:61]: just looking at this / thinking about it more, another way of fixing this is to change the function signature to:

func parseAndFilterCommandLine(allowedEnvVars []string, in *command_line.CommandLine) map[string]string {
...
}

and just modify in and return the envVarMap

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
Few small nits but this looks great overall and fixes a bunch of stuff that seems like we'd really want to fix! Thanks for setting this up (and bearing with me on the SA thing in general)

File: nogo_config.json (hunk 1 above)
Comment (tempoz) [reply nogo_config.json:55]: yeah, they probably should. there's no particular reason. I'll fix it.

File: server/build_event_protocol/event_parser/BUILD (hunk 3 above)
Comment (tempoz) [reply server/build_event_protocol/event_parser/BUILD:14]: lol yeah missed that.

File: server/build_event_protocol/event_parser/event_parser_test.go (hunk 4 above)
Comment (tempoz) [reply server/build_event_protocol/event_parser/event_parser_test.go:240]: I agree strongly that it is harder to read; however, as it stands, `invocation` contains copies of all these protos (some few of which it alters), so the old code was comparing against the protos that were the input to invocation and confirming nothing about the actual work done by the parser to populate the invocation. This way we instead actually check against the values populated in `invocation` by the parser, which is presumably the point of the test.

File: server/build_event_protocol/event_parser/event_parser.go (hunk 5 above)
Comment (tempoz) [reply server/build_event_protocol/event_parser/event_parser.go:61]: I much prefer that solution; if we don't need to preserve the original proto or prevent changes to the original proto from affecting the invocation (which i assume we don't since you're suggesting it), that will be much faster/more memory efficient/more elegant in general.

File: server/build_event_protocol/build_event_proxy/build_event_proxy.go (hunk 2 above)
Comment (tempoz) [reply server/build_event_protocol/build_event_proxy/build_event_proxy.go:83]: of course!

File: server/build_event_protocol/event_parser/event_parser_test.go (hunk 4 above)
Comment (tempoz) [reply server/build_event_protocol/event_parser/event_parser_test.go:240]: Even after implementing the change whereby we no longer make copies of the protos to store in `invocation`, we still need the test to ensure that the protos are being stored in `invocation` at all, so I think this change remains necessary even with the proposed signature change to `parseAndFilterCommandLine`

**Review by tylerwilliams — APPROVED** [review]
LGTM
